x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -0,0 +1,884 @@
+                "JWKSet contains RSA JWK in ["
+                    + configKeyJwkSetHmac
+                    + ", but there are no RSA signature algorithms. Signature algorithms are ["
+                    + String.join(",", allowedSignatureAlgorithms)
+                    + "]."
+            );
+        }
+        final int bitLengthRsa;
+        try {
+            bitLengthRsa = JwtUtil.computeBitLengthRsa(jwkRsa.toPublicKey());
+        } catch (Exception e) {
+            throw new SettingsException(
+                "JWKSet contains RSA JWK in [" + configKeyJwkSetHmac + "], but computing its bit length failed.",
+                e
+            );
+        }
+        if (bitLengthRsa < minBitLengthRsa) {
+            throw new SettingsException(
+                "JWKSet contains RSA JWK in ["
+                    + configKeyJwkSetHmac
+                    + "], but its bit length ["
+                    + bitLengthRsa
+                    + "] is smaller than minimum allowed by signature algorithms ["
+                    + String.join(",", allowedSignatureAlgorithmsRsa)
+                    + "]."
+            );
+        }
+    }
+
+    public static void validateJwkEc(final String configKeyJwkSetHmac, final ECKey jwkEc, final List<String> allowedSignatureAlgorithms) {
+        final List<String> allowedSignatureAlgorithmsEc = filter(
+            allowedSignatureAlgorithms,
+            JwtRealmSettings.SUPPORTED_SIGNATURE_ALGORITHMS_EC
+        );
+        final Set<Curve> allowedEcCurves = JwtUtil.computeAllowedCurvesEc(allowedSignatureAlgorithmsEc);
+        if (allowedEcCurves.isEmpty()) {
+            throw new SettingsException(
+                "JWKSet contains EC JWK in ["
+                    + configKeyJwkSetHmac
+                    + "], but there are no EC signature algorithms. Signature algorithms are ["
+                    + String.join(",", allowedSignatureAlgorithms)
+                    + "]."
+            );
+        }
+        if (allowedEcCurves.contains(jwkEc.getCurve()) == false) {
+            throw new SettingsException(
+                "JWKSet contains EC JWK in ["
+                    + configKeyJwkSetHmac
+                    + "], but there are no EC signature algorithms for that curve. EC signature algorithms are ["
+                    + String.join(",", allowedSignatureAlgorithmsEc)
+                    + "]. Curves are ["
+                    + allowedEcCurves
+                    + "]."
+            );
+        }
+    }
+
+    public static List<String> filter(final List<String> input, final List<String> filter) {
+        return filter.stream().filter(input::contains).collect(Collectors.toList());
+    }
+
+    public static void validateJwtAuthTime(final long allowedClockSkewSeconds, final Date now, final Date auth_time) throws Exception {
+        if (allowedClockSkewSeconds < 0L) {
+            throw new Exception("Invalid allowedClockSkewSeconds [" + allowedClockSkewSeconds + "] < 0.");
```

Comment:
JwtUtil.testValidateJwtIssuedAtTime() calls this and other methods to perform time-based checks. The methods are package private now. That test calls each method 7 or more times. I think it makes sense to leave them as is, otherwise I'd be looking at constructing full realm settings 7x4 just to run some boundary condition tests.